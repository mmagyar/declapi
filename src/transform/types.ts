import Ajv from 'ajv'
import { ObjectType } from 'yaschva'
import { Elastic, KeyValue } from '../DataDriverTypes.js'
import { HttpMethods, SearchTypes } from '../globalTypes.js'
import { dirname } from 'path'
import { fileURLToPath } from 'url'

export type AuthType = (string | {createdBy: boolean})[] | boolean
export type ManageableFields ={ createdBy?: boolean }

export type Contract = {
  name: string;
  type?: HttpMethods;
  authentication: AuthType;
  manageFields: ManageableFields;
  arguments: ObjectType;
  returns: ObjectType;
  description?: string;
};
export type CrudAuthAll = {
  get: AuthType;
  put: AuthType;
  post: string[] | boolean;
  delete: AuthType;
};

export type CrudAuthSome = {
  get: AuthType;
  modify: AuthType;
  delete?: AuthType;
};

export type CrudContract = {
  name: string;
  methods?: { get?:boolean, post?:boolean, put?: boolean, patch?:boolean, delete?:boolean},
  authentication: AuthType | CrudAuthAll | CrudAuthSome;
  manageFields?:ManageableFields;
  dataType: ObjectType;
  search?: SearchTypes;
  description?: string;
  preferredImplementation?: {type: 'elasticsearch'; index: string};
};

export type OutputSuccess = {
  name: string;
  authentication: AuthType;
  manageFields: ManageableFields;
  method: HttpMethods;
  arguments: ObjectType;
  returns: ObjectType;
  preferredImplementation?: Elastic | KeyValue;
  search?: SearchTypes;
};

export type Output =
  | {type: 'result'; key: string; results: OutputSuccess[]; errors?: undefined;}
  | {type: 'error'; key?: undefined; errors: (Ajv.ErrorObject | string)[] | string; results?: undefined;};

export const baseSchemaLocation = `${dirname(fileURLToPath(import.meta.url))}/../../schema/`
