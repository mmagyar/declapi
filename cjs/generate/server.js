"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.server = void 0;
const elastic_js_1 = __importDefault(require("./elastic.js"));
const common_js_1 = require("./common.js");
const kv_js_1 = require("./kv.js");
exports.server = (contracts) => {
    const valueDef = contracts.map(x => {
        let handle = 'undefined';
        if (x.preferredImplementation && x.preferredImplementation.type === 'elasticsearch') {
            const codeGenInput = x.method === 'GET'
                ? { method: x.method, search: x.search || 'idOnly' }
                : { method: x.method };
            handle = elastic_js_1.default(x.preferredImplementation, codeGenInput);
        }
        else if (x.preferredImplementation && x.preferredImplementation.type === 'key-value') {
            const codeGenInput = x.method === 'GET'
                ? { method: x.method, search: x.search || 'idOnly' }
                : { method: x.method };
            handle = kv_js_1.kvCodeGen(x.preferredImplementation, codeGenInput);
        }
        return `${common_js_1.name(x)}: {
          name: "${x.name}",
          manageFields: ${JSON.stringify(x.manageFields, undefined, 2)},
          authentication: ${JSON.stringify(x.authentication, undefined, 2)},
          type: "${x.method}",
          handle: ${handle},
          arguments: ${JSON.stringify(x.arguments)} ,
          returns: ${JSON.stringify(x.returns)}}`;
    }).join(',\n');
    const contractTypeList = contracts.map(x => `${common_js_1.name(x)}: ContractType<${common_js_1.name(x)}Argument, ${common_js_1.name(x)}Returns>`).join('\n');
    const elastic = contracts.find(x => { var _a; return ((_a = x.preferredImplementation) === null || _a === void 0 ? void 0 : _a.type) === 'elasticsearch'; });
    const kv = contracts.find(x => { var _a; return ((_a = x.preferredImplementation) === null || _a === void 0 ? void 0 : _a.type) === 'key-value'; });
    const elasticImport = elastic ? ', elastic' : '';
    const kvImport = kv ? ', kv' : '';
    const result = `/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType${elasticImport}${kvImport} } from "declarapi-runtime"

  ${common_js_1.typeDef(contracts)}
  export type ContractListType = {\n${contractTypeList}\n}\n
  export const contracts: ContractListType = {\n${valueDef}\n}\n`;
    return result;
};
exports.default = exports.server;
//# sourceMappingURL=server.js.map