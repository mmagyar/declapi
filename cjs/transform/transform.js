"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = void 0;
const crud_js_1 = require("./crud.js");
const single_js_1 = require("./single.js");
exports.transform = (contract) => __awaiter(void 0, void 0, void 0, function* () {
    const data = contract;
    if (!data.$schema) {
        return { type: 'error', errors: "Schema files must contain $schema that point to it's type" };
    }
    if (data.$schema.endsWith('singleContractSchema.json')) {
        return single_js_1.transform(data);
    }
    else if (data.$schema.endsWith('crudContractSchema.json')) {
        return crud_js_1.transform(data);
    }
    return { type: 'error', errors: `Unsupported schema for declaration: ${data.$schema}` };
});
exports.default = exports.transform;
//# sourceMappingURL=transform.js.map