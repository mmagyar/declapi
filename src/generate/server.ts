import { OutputSuccess } from '../transform/types.js'
import elasticCodeGen, { ElasticInputType } from './elastic.js'
import { name, typeDef } from './common.js'
import { kvCodeGen } from './kv.js'

export const server = (contracts: OutputSuccess[]): string => {
  const valueDef = contracts.map(x => {
    let handle = 'undefined'
    if (x.preferredImplementation && x.preferredImplementation.type === 'elasticsearch') {
      const codeGenInput:ElasticInputType = x.method === 'GET'
        ? { method: x.method, search: x.search || 'idOnly' }
        : { method: x.method }

      handle = elasticCodeGen(x.preferredImplementation, codeGenInput)
    } else if (x.preferredImplementation && x.preferredImplementation.type === 'key-value') {
      const codeGenInput:ElasticInputType = x.method === 'GET'
        ? { method: x.method, search: x.search || 'idOnly' }
        : { method: x.method }

      handle = kvCodeGen(x.preferredImplementation, codeGenInput)
    }

    return `${name(x)}: {
          name: "${x.name}",
          manageFields: ${JSON.stringify(x.manageFields, undefined, 2)},
          authentication: ${JSON.stringify(x.authentication, undefined, 2)},
          type: "${x.method}",
          handle: ${handle},
          arguments: ${JSON.stringify(x.arguments)} ,
          returns: ${JSON.stringify(x.returns)}}`
  }).join(',\n')

  const contractTypeList = contracts.map(x =>
    `${name(x)}: ContractType<${name(x)}Argument, ${name(x)}Returns>`).join('\n')

  const elastic = contracts.find(x => x.preferredImplementation?.type === 'elasticsearch')

  const kv = contracts.find(x => x.preferredImplementation?.type === 'key-value')

  const elasticImport = elastic ? ', elastic' : ''
  const kvImport = kv ? ', kv' : ''

  const result =
  `/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType${elasticImport}${kvImport} } from "declarapi-runtime"

  ${typeDef(contracts)}
  export type ContractListType = {\n${contractTypeList}\n}\n
  export const contracts: ContractListType = {\n${valueDef}\n}\n`

  return result
}

export default server
