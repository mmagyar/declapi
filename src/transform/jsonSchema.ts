import Ajv, { ErrorObject } from 'ajv'
export const jsonValidatorInit = () => {
  return new Ajv({
    loadSchema: async (uri: any) => require(uri.replace('node_modules/', ''))
  })
}

export type jsonValidationSuccess = {type:'success', errors? : ErrorObject[]};
export type jsonValidationError = {type:'error', errors : ErrorObject[]};
export type validateResult = jsonValidationSuccess | jsonValidationError;
export const isValidationError = (input:validateResult) : input is jsonValidationError => input && input.type === 'error'
export const validate = async (json: Object, data:any):
 Promise<validateResult> => {
  const validator = await jsonValidatorInit().compileAsync(json)
  const result = validator(data)

  const errors = validator.errors || []
  if (!result) return { type: 'error', errors }
  return { type: 'success' }
}
