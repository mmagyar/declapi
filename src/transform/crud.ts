import { ValueType, ObjectType, StringType } from 'yaschva'
import { map } from 'microtil'
import { validate, isValidationError } from './jsonSchema'
import {
  CrudContract,

  CrudAuthAll,
  CrudAuthSome, OutputSuccess, Output, baseSchemaLocation, ManageableFields
} from './types'
import {
  HttpMethods,
  SearchTypes
} from 'declarapi-runtime'

const contractOptions = (input: ValueType | ValueType[]): ValueType[] => {
  if (Array.isArray(input)) {
    if (input.some(x => x === '?')) { return input }

    return input.concat(['?'])
  }
  return [input, '?']
}

const searchToType =
  (idType: 'string'| StringType, dataType: ObjectType, search?: SearchTypes): ObjectType => {
    if (search === 'idOnly') {
      return { id: [idType, { $array: idType }, '?'] }
    } else if (search === 'textSearch') {
      return { search: ['string', '?'], id: [idType, { $array: idType }, '?'] }
    } else if (search === 'full') {
      return map(dataType, value => contractOptions(value))
    } else if (!search) {
      return {}
    }

    return search
  }

const checkIdField = (contract:CrudContract) :Output | false => {
  if (contract.dataType.id === undefined) {
    return {
      type: 'error',
      errors: 'id field does not exist in the data declaration'
    }
  }

  const idType: any = contract.dataType.id
  if (!(idType === 'string' || idType.$string)) {
    return {
      type: 'error',
      errors: 'Type of id field must be string'
    }
  }

  return false
}
const checkManageFields = (contract: CrudContract): Output|false => {
  for (const [key, value] of Object.entries(contract.manageFields || {})) {
    if (value) {
      const fieldType: any = contract.dataType[key]
      if (fieldType === undefined) {
        return {
          type: 'error',
          errors: `managed field "${key}" is not present on data type`
        }
      }

      if (!(fieldType === 'string' || fieldType.$string)) {
        return {
          type: 'error',
          errors: `managed field "${key}" must be a string, current type :${fieldType}`
        }
      }
    }
  }

  return false
}

const removeManaged = (args: ObjectType, manageFields?: ManageableFields):ObjectType => {
  const result = { ...args }
  for (const [key, value] of Object.entries(manageFields || {})) {
    if (value) {
      delete result[key]
    }
  }

  return result
}
const isCrudAuth = (tbd: any): tbd is CrudAuthAll => tbd.post !== undefined
const isCrudAuthSome = (tbd: any): tbd is CrudAuthSome => tbd.modify !== undefined
const transformForPost = (tbd: any) => Array.isArray(tbd) && tbd.find(x => x.createdBy) ? true : tbd
export const transform = async (data:CrudContract | any): Promise<Output> => {
  const valid = await validate(require(`${baseSchemaLocation}crudContractSchema.json`), data)
  if (isValidationError(valid)) return valid

  const contractData: CrudContract = data

  const au = contractData.authentication

  const auth = {
    GET: isCrudAuth(au) ? au.get : (isCrudAuthSome(au) ? au.get : au),
    POST: isCrudAuth(au) ? au.post : transformForPost((isCrudAuthSome(au) ? au.modify : au)),
    PUT: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    PATCH: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    DELETE: isCrudAuth(au) ? au.delete : (isCrudAuthSome(au) ? au.delete || au.modify : au)
  }

  const createOutput = (method: HttpMethods, args: ObjectType,
    returns: ObjectType = contractData.dataType): OutputSuccess => ({
    method,
    name: contractData.name,
    authentication: auth[method],
    manageFields: contractData.manageFields || {},
    search: contractData.search,
    preferredImplementation: contractData.preferredImplementation,
    arguments: args,
    returns
  })

  const returnArray = { $array: contractData.dataType }
  const output: OutputSuccess[] = []

  const errorWithId = checkIdField(contractData)
  if (errorWithId) return errorWithId

  const errorWithManageFields = checkManageFields(contractData)
  if (errorWithManageFields) return errorWithManageFields

  const idType: any = contractData.dataType.id
  if (contractData.methods?.get !== false) {
    const search = contractData.search
    output.push(createOutput('GET',
      searchToType(idType, contractData.dataType, search), returnArray))
  }

  if (contractData.methods?.post !== false) {
    const post = { ...contractData.dataType, id: [idType, '?'] }
    output.push(createOutput('POST', removeManaged(post, contractData.manageFields)))
  }

  if (contractData.methods?.put !== false) {
    output.push(createOutput('PUT', removeManaged(contractData.dataType, contractData.manageFields)))
  }

  if (contractData.methods?.patch !== false) {
    const patch: {[s: string]: ValueType | ValueType[];} =
    { ...map(contractData.dataType, contractOptions), id: idType }
    output.push(createOutput('PATCH', removeManaged(patch, contractData.manageFields)))
  }

  if (contractData.methods?.delete !== false) {
    const deleteIds: {[s: string]: ValueType[];} = { id: [idType, { $array: idType }] }
    output.push(createOutput('DELETE', deleteIds, returnArray))
  }

  return { type: 'result', key: contractData.name, results: output }
}

export default transform
