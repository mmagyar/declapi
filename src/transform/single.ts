import { validate, isValidationError } from './jsonSchema.js'
import { Contract, Output, baseSchemaLocation } from './types.js'
import { loadJSON } from '../util.js'

export const transform = async (data:Contract | any): Promise<Output> => {
  const valid = await validate(await loadJSON(`${baseSchemaLocation}singleContractSchema.json`), data)
  if (isValidationError(valid)) return valid
  const contractData: Contract = data
  return {
    type: 'result',
    key: contractData.name,
    results: [
      {
        name: contractData.name,
        authentication: contractData.authentication,
        manageFields: {},
        method: contractData.type || 'get',
        arguments: data.arguments,
        returns: data.returns
      }
    ]
  }
}

export default transform
