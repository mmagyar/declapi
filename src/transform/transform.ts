import { Output, CrudContract, Contract } from './types'
import { transform as crudTransform } from './crud'
import { transform as singleTransform } from './single'

export const transform = async (contract: CrudContract | Contract|string | object): Promise<Output> => {
  const data: any = typeof contract === 'string' ? require(contract) : contract
  if (!data.$schema) { return { type: 'error', errors: "Schema files must contain $schema that point to it's type" } }

  if (data.$schema.endsWith('singleContractSchema.json')) {
    return singleTransform(data)
  } else if (data.$schema.endsWith('crudContractSchema.json')) {
    return crudTransform(data)
  }
  return { type: 'error', errors: `Unsupported schema for declaration: ${data.$schema}` }
}

export default transform