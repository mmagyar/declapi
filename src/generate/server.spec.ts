import server from './server'
import { OutputSuccess } from '../transform/types'
describe('Generate typing and fetch function for server', () => {
  const singleExample = (): OutputSuccess[] => [
    {
      name: 'test',
      authentication: false,
      method: 'get',
      arguments: {
        myNumber: 'number'
      },
      returns: {}
    }
  ]

  const crudExample = (): OutputSuccess[] => [
    {
      method: 'get',
      name: 'test',
      authentication: false,
      arguments: { search: ['string', '?'], id: ['string', { $array: 'string' }, '?'] },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    },
    {
      method: 'post',
      name: 'test',
      authentication: false,
      arguments: { id: ['string', '?'], myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'put',
      name: 'test',
      authentication: false,
      arguments: { id: 'string', myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'patch',
      name: 'test',
      authentication: false,
      arguments: { id: 'string', myNumber: ['number', '?'] },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'delete',
      name: 'test',
      authentication: false,
      arguments: { id: ['string', { $array: 'string' }] },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    }
  ]
  it('Generates single example without an error that matches snapshot', () => {
    const result = server(singleExample())
    expect(result).toMatchSnapshot()
  })

  it('Generates crud example without an error that matches snapshot', () => {
    const result = server(crudExample())
    expect(result).toMatchSnapshot()
  })

  const crudElasticExample = (): OutputSuccess[] => [
    {
      method: 'get',
      name: 'test',
      authentication: false,
      search: 'textSearch',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: {
        search: ['string', '?'],
        id: ['string', { $array: 'string' }, '?']
      },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    },
    {
      method: 'post',
      name: 'test',
      authentication: false,
      search: 'textSearch',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: { id: ['string', '?'], myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'put',
      name: 'test',
      authentication: false,
      search: 'textSearch',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: { id: 'string', myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'patch',
      name: 'test',
      authentication: false,
      search: 'textSearch',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: { id: 'string', myNumber: ['number', '?'] },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'delete',
      name: 'test',
      authentication: false,
      search: 'textSearch',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: { id: ['string', { $array: 'string' }] },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    }
  ]

  it('generates elasticsearch implementation', () => {
    const result = server(crudElasticExample())
    expect(result).toMatchSnapshot()
  })

  const singleElasticExample = (): OutputSuccess[] => [
    {
      name: 'test',
      authentication: false,
      method: 'get',
      preferredImplementation: { type: 'elasticsearch', index: 'test' },
      arguments: {
        myNumber: 'number'
      },
      returns: {}
    }
  ]
  it('generates simple elasticsearch implementation, search defaults to idOnly get', () => {
    const result = server(singleElasticExample())
    expect(result).toMatchSnapshot()
  })

  it('full search is not implemented', () => {
    const data = singleElasticExample()
    data[0].search = 'full'
    expect(() => server(data)).toThrow('Parametric get not implemented yet')
  })

  it('custom search is not implemented', () => {
    const data = singleElasticExample()
    data[0].search = { littleKitty: 'string' }
    expect(() => server(data)).toThrow('Unsupported automatic elasticsearch methods: {"littleKitty":"string"}')
  })
})
