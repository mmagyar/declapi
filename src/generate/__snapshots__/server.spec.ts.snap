// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate typing and fetch function for server Generates crud example without an error that matches snapshot 1`] = `
"/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType } from \\"declarapi-runtime\\"

  /** test types for GET method **/
export type testGETArgument = { search?: string | undefined; id?: string | string[] | undefined } | undefined
export type testGETReturns = { id: string; myNumber: number }[]

/** test types for POST method **/
export type testPOSTArgument = { id?: string | undefined; myNumber: number }
export type testPOSTReturns = { id: string; myNumber: number }

/** test types for PUT method **/
export type testPUTArgument = { id: string; myNumber: number }
export type testPUTReturns = { id: string; myNumber: number }

/** test types for PATCH method **/
export type testPATCHArgument = { id: string; myNumber?: number | undefined }
export type testPATCHReturns = { id: string; myNumber: number }

/** test types for DELETE method **/
export type testDELETEArgument = { id: string | string[] }
export type testDELETEReturns = { id: string; myNumber: number }[]

  export type ContractListType = {
testGET: ContractType<testGETArgument, testGETReturns>
testPOST: ContractType<testPOSTArgument, testPOSTReturns>
testPUT: ContractType<testPUTArgument, testPUTReturns>
testPATCH: ContractType<testPATCHArgument, testPATCHReturns>
testDELETE: ContractType<testDELETEArgument, testDELETEReturns>
}

  export const contracts: ContractListType = {
testGET: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"GET\\",
          handle: undefined,
          arguments: {\\"search\\":[\\"string\\",\\"?\\"],\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"},\\"?\\"]} ,
          returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}},
testPOST: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"POST\\",
          handle: undefined,
          arguments: {\\"id\\":[\\"string\\",\\"?\\"],\\"myNumber\\":\\"number\\"} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testPUT: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"PUT\\",
          handle: undefined,
          arguments: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testPATCH: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"PATCH\\",
          handle: undefined,
          arguments: {\\"id\\":\\"string\\",\\"myNumber\\":[\\"number\\",\\"?\\"]} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testDELETE: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"DELETE\\",
          handle: undefined,
          arguments: {\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"}]} ,
          returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}}
}
"
`;

exports[`Generate typing and fetch function for server Generates single example without an error that matches snapshot 1`] = `
"/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType } from \\"declarapi-runtime\\"

  /** test types for GET method **/
export type testGETArgument = { myNumber: number }
export type testGETReturns = {  } | undefined

  export type ContractListType = {
testGET: ContractType<testGETArgument, testGETReturns>
}

  export const contracts: ContractListType = {
testGET: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"GET\\",
          handle: undefined,
          arguments: {\\"myNumber\\":\\"number\\"} ,
          returns: {}}
}
"
`;

exports[`Generate typing and fetch function for server generates elasticsearch implementation 1`] = `
"/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType, elastic } from \\"declarapi-runtime\\"

  /** test types for GET method **/
export type testGETArgument = { search?: string | undefined; id?: string | string[] | undefined } | undefined
export type testGETReturns = { id: string; myNumber: number }[]

/** test types for POST method **/
export type testPOSTArgument = { id?: string | undefined; myNumber: number }
export type testPOSTReturns = { id: string; myNumber: number }

/** test types for PUT method **/
export type testPUTArgument = { id: string; myNumber: number }
export type testPUTReturns = { id: string; myNumber: number }

/** test types for PATCH method **/
export type testPATCHArgument = { id: string; myNumber?: number | undefined }
export type testPATCHReturns = { id: string; myNumber: number }

/** test types for DELETE method **/
export type testDELETEArgument = { id: string | string[] }
export type testDELETEReturns = { id: string; myNumber: number }[]

  export type ContractListType = {
testGET: ContractType<testGETArgument, testGETReturns>
testPOST: ContractType<testPOSTArgument, testPOSTReturns>
testPUT: ContractType<testPUTArgument, testPUTReturns>
testPATCH: ContractType<testPATCHArgument, testPATCHReturns>
testDELETE: ContractType<testDELETEArgument, testDELETEReturns>
}

  export const contracts: ContractListType = {
testGET: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"GET\\",
          handle: (input, auth, contract) => elastic.get(\\"test\\", contract, auth, input && input.id, input && input.search),
          arguments: {\\"search\\":[\\"string\\",\\"?\\"],\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"},\\"?\\"]} ,
          returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}},
testPOST: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"POST\\",
          handle: (input, auth, contract) => elastic.post(\\"test\\", contract, auth, input),
          arguments: {\\"id\\":[\\"string\\",\\"?\\"],\\"myNumber\\":\\"number\\"} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testPUT: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"PUT\\",
          handle: (input, auth, contract) => elastic.put(\\"test\\", contract, auth, input, input.id),
          arguments: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testPATCH: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"PATCH\\",
          handle: (input, auth, contract) => elastic.patch(\\"test\\", contract, auth, input, input.id),
          arguments: {\\"id\\":\\"string\\",\\"myNumber\\":[\\"number\\",\\"?\\"]} ,
          returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},
testDELETE: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"DELETE\\",
          handle: (input, auth, contract) => elastic.del(\\"test\\", contract, auth, input.id),
          arguments: {\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"}]} ,
          returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}}
}
"
`;

exports[`Generate typing and fetch function for server generates simple elasticsearch implementation, search defaults to idOnly get 1`] = `
"/**********************************************
   DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/
  import { ContractType, elastic } from \\"declarapi-runtime\\"

  /** test types for GET method **/
export type testGETArgument = { myNumber: number }
export type testGETReturns = {  } | undefined

  export type ContractListType = {
testGET: ContractType<testGETArgument, testGETReturns>
}

  export const contracts: ContractListType = {
testGET: {
          name: \\"test\\",
          manageFields: {},
          authentication: false,
          type: \\"GET\\",
          handle: (input, auth, contract) => elastic.get(\\"test\\", contract, auth, input && input.id),
          arguments: {\\"myNumber\\":\\"number\\"} ,
          returns: {}}
}
"
`;
